### PR TITLE
Update to use DSCResource.test Shared Code - Fixes #177

### LINKED ISSUE #177: Error in AppVeyor CI because NUPKG is not produced
An error is occurring at the end of the AppVeyor CI build process. It appears that the build is not producing the Nuget package:

image

### PR DESCRIPTION
**Pull request by PlagueHO**
This PR changes:
- Converted to use AppVeyor.psm1 in DSCResource.Tests repository.
- Converted to use Example and Markdown tests in DSCResource.Tests repository.

This also fixes #177

---
This change is [Reviewable](https://reviewable.io/reviews/powershell/xnetworking/186)

### CODE REVIEW

**Comment by PlagueHO** [comment]
@tysonjhayes - I have updated this PR with the new MetaTestOptIn.json file to opt in to the Markdown and Example tests. Everything should be good to go now.

File: .MetaTestOptIn.json [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[
+        "Common Tests - Validate Markdown Files",
+        "Common Tests - Validate Example Files"
+]
```
Comment (TravisEz13) [inline .MetaTestOptIn.json]: missing newline at end of file

Comment (PlagueHO) [inline .MetaTestOptIn.json]: Fixed.

**Comment by johlju** [comment]
Reviewed 10 of 11 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[appveyor.yml, line 12 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqAhPhRlbKIOFPBYnV:-KcqAhPhRlbKIOFPBYnW:b-3r168j) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/appveyor.yml#L12)):*
> ```YAML
>         $mainModuleFolder = "Modules\$moduleName"
>         $harnessModulePath = "Tests\$($moduleName).TestHarness.psm1"
>         $harnessFunctionName = "Invoke-$($moduleName)Test"
> ```

Question: Couldn't we make this function generic in the future, the same name for all resources that uses Harness? Like `Invoke-HarnessTest` or just plain `Invoke-Test`.
Or is there a point to having the name in the function like`Invoke-xNetworkingTest`?

---

*[.vscode/RunAllTests.ps1, line 2 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqD2uxc1rMKO6WTMQm:-KcqD2uxc1rMKO6WTMQn:bk33ud2) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/.vscode/RunAllTests.ps1#L2)):*
> ```PowerShell
> Import-Module (Join-Path $PSScriptRoot "..\Tests\xNetworking.TestHarness.psm1" -Resolve)
> $dscTestsPath = Join-Path -Path $PSScriptRoot `
> ```

This one expects the DscResource.Tests to already been cloned? What if it isn't?

---

*[Tests/xNetworking.TestHarness.psm1, line 6 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqELwOezXrRaYJxbKJ:-KcqELwOezXrRaYJxbKK:b-cml0db) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/Tests/xNetworking.TestHarness.psm1#L6)):*
> ```PowerShell
>     param
>     (
>         [System.String] $TestResultsFile,
> ```

Could we format these two parameters like

```
[Parameter()] 
[System.String]
$TestResultsFile
```

---

*[Tests/xNetworking.TestHarness.psm1, line 13 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqEbfibrXB2r9IhDc8:-KcqEbfibrXB2r9IhDc9:bjz3lmy) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/Tests/xNetworking.TestHarness.psm1#L13)):*
> ```PowerShell
>     Write-Verbose -Message 'Commencing all xNetworking tests'
> 
>     $repoDir = Join-Path -Path $PSScriptRoot -ChildPath "..\" -Resolve
> ```

Single quotes

---

*[Tests/xNetworking.TestHarness.psm1, line 17 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqErkhJTgmFWbbjEKh:-KcqErkhJTgmFWbbjEKi:b6nnrp7) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/Tests/xNetworking.TestHarness.psm1#L17)):*
> ```PowerShell
>     $testCoverageFiles = @()
>     Get-ChildItem -Path "$repoDir\modules\xNetworking\DSCResources\**\*.psm1" -Recurse | ForEach-Object {
>         if ($_.FullName -notlike '*\DSCResource.Tests\*') {
> ```

Could this happen? Shouldn't DSCResource.Tests always end up outside DSCResources folder? If not, then code in RunAllTests.ps1 won't work. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xnetworking/186)*

**Comment by PlagueHO** [comment]
@johlju - regarding the name of the name of the harness invoke call. I fully agree - a consistent name would be much better here. I can change xNetworking and xStorage.

The question will be down to @BrianFarnhill: Brian are you ok with changing the name of your Invoke-SharePointDscTests function to a generic name?

I'll look at the rest of the comments tomorrow! Thanks for reviewing @johlju

**Comment by PlagueHO** [comment]
@johlju - I've been thinking about this and I think we should default the name of the Harness function to ```Invoke-HarnessTest```. If @BrianFarnhill wants to use the alternate name then we can just allow the parameter to be passed. That way we can use the default in all our repos but still support SharePointDsc.

I'll submit a PR with the changes to DSCREsource.tests and also to xNetworking and xStorage.

**Comment by PlagueHO** [comment]
Review status: 8 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[appveyor.yml, line 12 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqAhPhRlbKIOFPBYnV:-KcvSXL8dayP0-RwXFRo:b-g2hkyv) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/appveyor.yml#L12)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Question: Couldn't we make this function generic in the future, the same name for all resources that uses Harness? Like `Invoke-HarnessTest` or just plain `Invoke-Test`.
Or is there a point to having the name in the function like`Invoke-xNetworkingTest`?
</blockquote></details>

Done. I've submitted the PR to DSCResource.Tests. Once that is through this will start working. Note, I used:
```Invoke-TestHarness```.
```


---

*[.vscode/RunAllTests.ps1, line 2 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqD2uxc1rMKO6WTMQm:-KcvSk4X31vwM9oiMP50:b-896fix) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/.vscode/RunAllTests.ps1#L2)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

This one expects the DscResource.Tests to already been cloned? What if it isn't?
</blockquote></details>

Done.

---

*[Tests/TestHarness.psm1, line 6 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqELwOezXrRaYJxbKJ:-KcvSm4HHImogXM0cK0E:b-896fix) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/Tests/TestHarness.psm1#L6)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could we format these two parameters like

```
[Parameter()] 
[System.String]
$TestResultsFile
```
</blockquote></details>

Done.

---

*[Tests/TestHarness.psm1, line 13 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqEbfibrXB2r9IhDc8:-KcvSn1s31L72vLo5JSN:b-896fix) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/Tests/TestHarness.psm1#L13)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Single quotes
</blockquote></details>

Done.

---

*[Tests/TestHarness.psm1, line 17 at r2](https://reviewable.io:443/reviews/powershell/xnetworking/186#-KcqErkhJTgmFWbbjEKh:-KcvSnZ0iyl2-ToD1t3Q:b-2gd7o5) ([raw file](https://github.com/powershell/xnetworking/blob/b77c26836170d4307b11f054b5020f1d2b067959/Tests/TestHarness.psm1#L17)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Could this happen? Shouldn't DSCResource.Tests always end up outside DSCResources folder? If not, then code in RunAllTests.ps1 won't work. :)
</blockquote></details>

It can happen. You can still run individual sets of tests on each repo by changing to the folder containing the specific tests and executing them which will clone into the DSCResources folder. So I think is better to be safe than sorry here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xnetworking/186)*
<!-- Sent from Reviewable.io -->

**Comment by TravisEz13** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xnetworking/186)*

**Comment by TravisEz13** [comment]
95% code coverage impressive.  You can explore the code coverage per file [in the branch view.](https://codecov.io/gh/PowerShell/xNetworking/tree/Use-AppVeyor-Module/Modules/xNetworking/DSCResources)

**Comment by johlju** [comment]
Reviewed 3 of 4 files at r3, 4 of 4 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[.codecov.yml, line 20 at r4](https://reviewable.io:443/reviews/powershell/xnetworking/186#-Kd1Yib5LfOT_AJsLd14:-Kd1Yib5LfOT_AJsLd15:b-lxylmn) ([raw file](https://github.com/powershell/xnetworking/blob/3b81c69d773741453b270554cf4e129d6568131e/.codecov.yml#L20)):*
> ```YAML
>       default:
>         # Set the overall project code coverage requirement to 70%
>         target: 70
> ```

Minor: Maybe raise target on this repo? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xnetworking/186)*

**Comment by TravisEz13** [comment]
Our HQRM guidelines are for 70%.  There is evidence that higher targets are not always better (although most of those assumed integration tests were not included in the goal.)  I would not raise it above 80.

**Comment by johlju** [comment]
@TravisEz13 I'm good with 70% if that is the guideline. But I thought we could use this check to fail if a new resource got the limit under 95% (in this repos case).

**Comment by johlju** [comment]
Aha I see! I did not see the `threshold` value bit in your text until now. Then I understand how it works! Yes maybe it would be good to lower it. Maybe @PlagueHO need to trim the value in.

**Comment by johlju** [comment]
:lgtm:

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xnetworking/186#-:-Kd1bs23iEVaFwlUe6kz:bnfp4nl)*

**Comment by PlagueHO** [comment]
@johlju , @TravisEz13 - I'm ok with lowering the threshold to a lower value. We could try say 2% and see how that goes. @tysonjhayes ?

**Comment by TravisEz13** [comment]
@PlagueHO That sounds good.  We should just remember to treat it with reason.  It's all very new.

**Comment by tysonjhayes** [comment]
Wouldn't lowering the threshold value increase the amount of code coverage needed? So new resources would have to meet the 95+/- 2% code coverage needed before they could be submitted? Seems like a tall order.

**Comment by TravisEz13** [comment]
Threshold says that a PR that is -<threshold>% will be marked as failing.  So yes, it would increase the requirement.  But don't we want to decrease code coverage regression?  You can still merge if these fail as well.

**Comment by BrianFarnhill** [comment]
@PlagueHO Sorry for being late to the party here, had a busy week. To answer the question of me, nah I don't mind using a generic name at all for the test suit. I'm happy to play nice with others on this one, no biggy at all.

**Comment by PlagueHO** [comment]
No worries @BrianFarnhill mate! I actually changed it so you can use a generic name or override it with your own one by specifying parameter values (AppVeyor.yml is smaller if you use generic).

**Comment by PlagueHO** [comment]
@tysonjhayes , @TravisEz13 - perhaps we should leave xNetworking threshold as is for now and see how things go. I guess the whole codecov thing is a bit new (at least to these repos) so we might want to monitor for a bit.

**Comment by TravisEz13** [comment]
@PlagueHO It is probably better to be a bit cautious in the beginning.  This is why I set 70% project requirement and 5% threshold.  It seemed to give a good amount of wiggle room while telling people when they met the project goals.

**Comment by PlagueHO** [comment]
Ok - I think we should just leave it as is and merge it. @TravisEz13, @tysonjhayes - are you happy for me to merge?

**Comment by PlagueHO** [comment]
@Johlju, I think there is still one unreviewed file...

---

Review status: 12 of 13 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/powershell/xnetworking/186#-:-KdJVQy7x92LspCOiVse:b-10qqje)*

**Pull request merged**